framework/py/flwr/serverapp/strategy/dp_adaptive_clipping.py

```diff
@@ -0,0 +1,343 @@
+    num_sampled_clients : int
+        The number of clients that are sampled on each round.
+    initial_clipping_norm : float
+        The initial value of clipping norm. Defaults to 0.1.
+        Andrew et al. recommends to set to 0.1.
+    target_clipped_quantile : float
+        The desired quantile of updates which should be clipped. Defaults to 0.5.
+    clip_norm_lr : float
+        The learning rate for the clipping norm adaptation. Defaults to 0.2.
+        Andrew et al. recommends to set to 0.2.
+    clipped_count_stddev : float
+        The stddev of the noise added to the count of
+        updates currently below the estimate.
+        Andrew et al. recommends to set to `expected_num_records/20`
+
+    Examples
+    --------
+    Create a strategy::
+
+        strategy = fl.serverapp.FedAvg(...)
+
+    Wrap the strategy with the `DifferentialPrivacyClientSideAdaptiveClipping` wrapper::
+
+        dp_strategy = DifferentialPrivacyClientSideAdaptiveClipping(
+            strategy, cfg.noise_multiplier, cfg.num_sampled_clients, ...
+        )
+
+    On the client, add the `adaptiveclipping_mod` to the client-side mods::
+
+        app = fl.client.ClientApp(mods=[adaptiveclipping_mod])
+    """
+
+    def __repr__(self) -> str:
+        """Compute a string representation of the strategy."""
+        return "Differential Privacy Strategy Wrapper (Client-Side Adaptive Clipping)"
+
+    def configure_train(
+        self, server_round: int, arrays: ArrayRecord, config: ConfigRecord, grid: Grid
+    ) -> Iterable[Message]:
+        """Configure the next round of training."""
+        config[KEY_CLIPPING_NORM] = self.clipping_norm
+        return self.strategy.configure_train(server_round, arrays, config, grid)
+
+    def aggregate_train(
+        self, server_round: int, replies: Iterable[Message]
+    ) -> tuple[Optional[ArrayRecord], Optional[MetricRecord]]:
+        """Aggregate ArrayRecords and MetricRecords in the received Messages."""
+        if not self._validate_replies(replies):
+            return None, None
+
+        replies_list = list(replies)
+        aggregated_arrays, aggregated_metrics = self.strategy.aggregate_train(
+            server_round, replies_list
+        )
+
+        self._update_clip_norm_from_replies(replies_list)
+
+        if aggregated_arrays:
+            aggregated_arrays = self._add_noise_to_aggregated_arrays(aggregated_arrays)
+
+        return aggregated_arrays, aggregated_metrics
+
+    def _update_clip_norm_from_replies(self, replies: list[Message]) -> None:
+        total = len(replies)
+        clipped_count = 0
+
+        for msg in replies:
+            found = False
+            for _, mrec in msg.content.metric_records.items():
+                if KEY_NORM_BIT in mrec:
+                    clipped_count += int(bool(mrec[KEY_NORM_BIT]))
+                    found = True
+                    break
+            if found:
+                continue
```

Comment:
should we throw an error and hault the aggregation process if `KEY_NORM_BIT` is not found? seeing also the lines after this comment what i'm thinking we could do is add a stage at the top of `aggregate_train` that checks if `KEY_NORM_BIT` is found in the single `MetricRecord` arriving in the replies. if not, we can throw a `serverapp.exception.AggregationError` in this way we no longer need to do all these check  later on (since these keys are guaranteed to be present). wdyt?